### PR TITLE
Extend StatefulHeaderFormatter to allow forwarding HTTP1 reason phrase

### PR DESCRIPTION
**Pull request by syhpoon**
Commit Message: http: extend StatefulHeaderFormatter to allow forwarding HTTP1 reason phrase
Additional Description: In some cases the upstream can produce a non-default reason phrase when using HTTP1. For example AWS S3 can return `503 Slow Down` in certain cases. StatefuleHeaderFormatter can now take an optional flag to enable forwarding non-default HTTP1 reason phrases.
Risk Level: Medium
Testing: unit testing, integration testing
Docs Changes: release notes
Release Notes: http: added support for forwarding HTTP1 reason phrase
Platform Specific Features: N/A

### CODE REVIEW

**Comment by mattklein123** [comment]
Thank you for the contribution. I'm not sure if you are the same person that I was discussing this on Slack with (https://envoyproxy.slack.com/archives/C78HA81DH/p1634828677000900), but I think we decided that the best and least intrusive way of supporting this is to add this into the stateful header formatter system that allows remembering the case of H1 headers during proxy. I think adding remembering the reason phrase to this would be pretty simple and avoid polluting the header map with H1 specific concepts. Can we try to implement it that way? Thank you.

/wait

**Comment by syhpoon** [comment]
Yes, that was me discussing this issue on Slack.
There are couple of issues with the stateful formatter approach that didn't seem very elegant to me:

* Let's say I augment `StatefulHeaderKeyFormatter` interface with a pair of new methods: `set/getReasonPhrase`, then I either need to create a brand-new formatter class which `only` implements the new methods (and thus leaving the formatter ones hanging), or extend the existing `PreserveCaseHeaderFormatter` one, in which case the class name becomes confusing since not only can it format headers to preserve case but also (unrelatedly) manipulate reason phrase. Whatever route I go, it seems like current and future implementations of the interface will only care about either formatting headers with some custom logic OR preserving the reason phrase (and I'm having a hard time figuring out different type of implementations for this feature).
* In either of the cases above, the configuration becomes non-stackable, since there can be only one `stateful_formatter` config parameter, and one wouldn't be able to enable `both` preserving header casing and proxying reason phrase, especially with more stateful formatters implemented in the future. This is another sign for me that these two features do not belong in the same interface.

Current implementation make this an HTTP1-specific option. If the option is not set (default), there's no performance penalty as the parser callback is not even called. The only addition is one new class member field, and yes, I agree, it's not ideal to have a http1-specific concept in generic class, but I was kind hoping that folks much more familiar with the codebase could advise on a better way to achieve this, if this is considered a major concern.

**Comment by mattklein123** [comment]
This feature is a very niche ask, similar to preserving the case of H1 headers. I personally would not overthink this. I think it would be fine to just add an option to https://github.com/envoyproxy/envoy/blob/main/api/envoy/extensions/http/header_formatters/preserve_case/v3/preserve_case.proto about whether to preserve the reason phrase, or, I think it woud probably be fine to just do it always when using this formatter.

Then I think it's fine to extend the interface to set/get the reason phrase.

**Comment by syhpoon** [comment]
Okay, then a follow-up question: would you expect for `StatefulHeaderKeyFormatter` interface to ever get another implementation (in addition to `PreserveCaseFormatter`)? If not, then sure, it should be fine to add new methods and implement them in the existing class, but if there's a chance there will be other implementations (and from the way it's currently structured it surely seems like this was the plan), then it all becomes pretty messy, since all future implementations would have to override set/getReasonPhrase methods while it wouldn't make a lot of sense to have many implementations of this logic, plus there won't be an easy way to enable multiple key formatters, but I already mentioned that.

Basically this is my main concern, but if you're saying it's okay to go this way, I can definitely tweak my implementation accordingly.

**Comment by mattklein123** [comment]
> then it all becomes pretty messy, since all future implementations would have to override set/getReasonPhrase methods while it wouldn't make a lot of sense to have many implementations of this logic, plus there won't be an easy way to enable multiple key formatters, but I already mentioned that.

I don't think it's really that messy. The interface methods could be a nop in a different implementation. As long as we clearly document that the PreserveCaseFormatter does this, I think it's fine. I agree the name is now not great, but so it goes. I just don't think it's worth leaking the implementation details of this very nice feature into the rest of the codebase.

> Basically this is my main concern, but if you're saying it's okay to go this way, I can definitely tweak my implementation accordingly.

Let's get another opinion from @alyssawilk on this. @alyssawilk for preserving the H1 reason phrase, do you think we should bolt it on to the stateful header formatter code or build it more directly in?

**Comment by alyssawilk** [comment]
I'm less familiar with the stateful header formatters so largely inclined to trust you on this, but naming aside the formatters largely exist to handle "metadata" for how we want to serialize HTTP/1.1 requests to the wire, so I'm totally fine jamming the reason phrase in there too.

File: api/envoy/extensions/http/header_formatters/preserve_case/v3/preserve_case.proto
Comment (mattklein123) [inline api/envoy/extensions/http/header_formatters/preserve_case/v3/preserve_case.proto:19]: Add a release note for this please

File: envoy/http/header_formatter.h
Comment (mattklein123) [inline envoy/http/header_formatter.h:40]: Pass absl::string_view

File: source/common/http/http1/legacy_parser_impl.cc
Comment (mattklein123) [inline source/common/http/http1/legacy_parser_impl.cc:41]: I would just always parse status, the cost is low.

File: source/extensions/http/header_formatters/preserve_case/preserve_case_formatter.h
Comment (mattklein123) [inline source/extensions/http/header_formatters/preserve_case/preserve_case_formatter.h:27]: This needs to be std::string or you will be referencing data that is gone.

**Review by mattklein123 — CHANGES_REQUESTED** [review]
Looks generally on the right track, some comments to get you started, thanks.

/wait

**Comment by jmarantz** [comment]
@syhpoon please avoid force-pushing in Envoy Github as it disrupts the comment-flow. Just do normal merges and pushes.

File: api/envoy/extensions/http/header_formatters/preserve_case/v3/preserve_case.proto
Comment (syhpoon) [reply api/envoy/extensions/http/header_formatters/preserve_case/v3/preserve_case.proto:19]: Added.

File: envoy/http/header_formatter.h
Comment (syhpoon) [reply envoy/http/header_formatter.h]: Done.

File: source/common/http/http1/legacy_parser_impl.cc
Comment (syhpoon) [reply source/common/http/http1/legacy_parser_impl.cc]: Removed the flag.

File: source/extensions/http/header_formatters/preserve_case/preserve_case_formatter.h
Comment (syhpoon) [reply source/extensions/http/header_formatters/preserve_case/preserve_case_formatter.h]: Fixed.

**Comment by syhpoon** [comment]
@mattklein123 all the comments should have been addressed, would appreciate another :eyes:

**Comment by alyssawilk** [comment]
Apologies for delay but Matt's out this week and where normally we'd reassign, over half the maintainers are out as well so thanks for your patience on unusual review lag :-)

**Comment by syhpoon** [comment]
No worries! Thanks, Alyssa for letting me know!

**Comment by mattklein123** [comment]
@syhpoon friendly request to please not force push. It makes reviews much harder. Thank you!

**Comment by syhpoon** [comment]
Sure, but there were no new comments and previous CI run failed, it didn't look the force-push would mess anything.

File: test/extensions/http/header_formatters/preserve_case/preserve_case_formatter_test.cc
Comment (mattklein123) [inline test/extensions/http/header_formatters/preserve_case/preserve_case_formatter_test.cc:12]: Can you verify that when this is false, and you set a reason phrase, you don't get a reason phrase back?

File: test/extensions/http/header_formatters/preserve_case/preserve_case_formatter_reason_phrase_integration_test.cc
Comment (mattklein123) [inline test/extensions/http/header_formatters/preserve_case/preserve_case_formatter_reason_phrase_integration_test.cc:50]: Can you add an off test also where when explicitly disabled you get back the stock phrase?

**Review by mattklein123 — CHANGES_REQUESTED** [review]
Thanks, looks awesome. Just 2 small test comments then we can ship!

/wait

**Review by mattklein123 — COMMENTED** [review]
LGTM but needs format fix, thanks.

/wait

File: test/extensions/http/header_formatters/preserve_case/preserve_case_formatter_test.cc
Comment (syhpoon) [reply test/extensions/http/header_formatters/preserve_case/preserve_case_formatter_test.cc:12]: Added another test case for this.

File: test/extensions/http/header_formatters/preserve_case/preserve_case_formatter_reason_phrase_integration_test.cc
Comment (syhpoon) [reply test/extensions/http/header_formatters/preserve_case/preserve_case_formatter_reason_phrase_integration_test.cc]: Done!

**Comment by syhpoon** [comment]
Thanks @mattklein123, should be good to go!

**Comment by syhpoon** [comment]
Thanks, Matt!

**Pull request merged**
